README.md
<!-- markdownlint-disable MD033-->
# **Implode-CSS** 🌋💣

Npm CLI package that removes all unused CSS *classes* and *ids* from your project.

---

[![Version](https://badge.fury.io/js/%40josee9988%2Fimplode-css.svg)](https://www.npmjs.com/package/@josee9988/implode-css)
[![Downloads](https://img.shields.io/npm/dt/@josee9988/implode-css.svg)](https://www.npmjs.com/package/@josee9988/implode-css)

---

## **Goal of the package** 🎯

To remove **useless/unused** CSS selectors from a project to increase **performance** and **readability**.

If you create a project from scratch or use templates there is a *high chance of creating CSS selectors that you will finally not use* and will be downloaded by the user and will make the **page performance slower**.

With `implodeCss` you will be able to **detect** and/or **remove** these unused CSS.

---

## **Installation** 🔩⚙

``` bash
sudo npm install -g @josee9988/implode-css
```

---

## **Examples of use** 📐🛡

- If you **do not know** how to use the package, show the **manual** by using:

    ``` bash
    implodeCss -h # to show the manual
    ```

- To be **prompted by the package**, do not use arguments, *we will help you*:

    ``` bash
    implodeCss # it will guide you
    ```

- Specify **folder** (actual / absolute path):

    ``` bash
    implodeCss . # actual folder, you must use always absolute paths but with the dot
    ```

    ``` bash
    implodeCss /home/user/project # folder /home/user/project
    ```

- **Audit** (detect but do not do any action):

    ``` bash
    implodeCss . -a # audit actual package
    ```

- **Fix** (remove all unused CSS selectors):

    ``` bash
    implodeCss . -f # audit actual package
    ```

- **Ignore folders**:

    ``` bash
    implodeCss . -a -i src,data  # audit actual package and do not look into folders: /src/, /data/
    ```

- Specify **port** (default = 4949):

    ``` bash
    implodeCss . -a -p 8080 # audit actual package and create the server on port 4949
    ```

---

## **Screenshots** 📸 😁

- **Auditing** the code with: `implodeCss . -a`

<img src="https://i.imgur.com/AnTvuHb.png" alt="auditing code" title="auditing code" style="border-radius:15px; box-shadow: 6px 6px  #282829; max-height: 500px; max-width:500px;margin-left: auto; margin-right:auto;display: block;margin-left: auto;margin-right:auto;width:50%;"/>

- **Table** with all the unused selectors and more information.

<img src="https://i.imgur.com/vkXgq70.png" alt="table with unused selectors" title="table with unused selectors" style="border-radius:15px; box-shadow: 6px 6px  #282829; max-height: 500px; max-width:500px;margin-left: auto; margin-right:auto;display: block;margin-left: auto;margin-right:auto;width:50%;"/>

- **Help** command: `implodeCss -h`

<img src="https://i.imgur.com/dVof1Rx.png" alt="help command" title="help command" style="border-radius:15px; box-shadow: 6px 6px  #282829; max-height: 500px; max-width:500px;margin-left: auto; margin-right:auto;display: block;margin-left: auto;margin-right:auto;width:50%;"/>

- If you do not specify a folder...

<img src="https://i.imgur.com/HJj8OmH.png" alt="folder not specified" title="folder not specified" style="border-radius:15px; box-shadow: 6px 6px  #282829; max-height: 500px; max-width:500px;margin-left: auto; margin-right:auto;display: block;margin-left: auto;margin-right:auto;width:50%;"/>

- If you do not specify the option to perform...

<img src="https://i.imgur.com/l67EvSs.png" alt="option not specified" title="option not specified" style="border-radius:15px; box-shadow: 6px 6px  #282829; max-height: 500px; max-width:500px;margin-left: auto; margin-right:auto;display: block;margin-left: auto;margin-right:auto;width:50%;"/>

---

## **Error codes** ❌

<details>
<summary>Click to see all the <b>error</b> and <b>informational</b> codes and it's explanation</summary>

### **200**

- Exists without errors.

### **201**

- Exit without an expected code, this code should not appear at all.

### **401**

- Bad arguments passed by the user. Some of the arguments used are not expected.

### **402**

- The path (most of the times the main folder) doesn't have read permissions, so the program could not read your files.

### **403** [Not used]

- The path doesn't have write permissions (Not used anywhere yet. see 405 instead).

### **404**

- Found 0 CSS files or 0 HTML files. The user should provide atleast 1 CSS files and 1 HTML file.

### **405**

- The local temporary file doesn't have write permissions so the page with the results won't get any data.

### **406**

- A path inside the main folder doesn't have read or write permissions (not specified).

### **500**

- Unknown error, the package specified an error that doesn't exist (should not appear, if so, please contact us and tell us how).

### **501**

- The selected port couldn't be opened.

### **502**

- Unexpected error (should not appear, if so, please contact us and tell us how).

</details>

---

### Did you enjoyed the package? Help us raise these numbers up 🥰 🎉

[![Github followers](https://img.shields.io/github/followers/Josee9988.svg?style=social)](#did-you-enjoyed-the-package-help-us-raise-these-numbers-up--)
[![Github stars](https://img.shields.io/github/stars/Josee9988/Implode-CSS.svg?style=social)](#did-you-enjoyed-the-package-help-us-raise-these-numbers-up--)
[![Github watchers](https://img.shields.io/github/watchers/Josee9988/Implode-CSS.svg?style=social)](#did-you-enjoyed-the-package-help-us-raise-these-numbers-up--)
[![Github forks](https://img.shields.io/github/forks/Josee9988/Implode-CSS.svg?style=social)](#did-you-enjoyed-the-package-help-us-raise-these-numbers-up--)

Also check the [npm page with the package](https://www.npmjs.com/package/@josee9988/implode-css)

---

> ⚠️Remember that this tool does not guarantee a 100% effectiveness and may have some issue at some point. Use it at your own risk and always do backups of your code.⚠️

*Made with a lot of ❤️❤️ by **[@Josee9988](https://github.com/Josee9988)***


CHANGELOG.md
<!-- markdownlint-disable MD024-->
# **Change Log** 📜📝

All notable changes to the "**ImplodeCss**" package will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

---

## [**1.3.2**] - 2020-01-16

### Fixed

* Better error handling in the HTML viewpage.
* Problem that made the Fix option not work at all.

### Added

* WebPage uses TypeScript to create the controller which gathers the data and displays it.
* Npm node script 'npm run build' must be runned before doing any pushes to the repo as it removes all the data from the temp file (data.js) and transpiles the TypeScript into JavaScript.
* Folder .vscode to be uploaded to GitHub.
* Extension recommendations.

## [**1.3.1**] - 2019-12-23

### Added

* Information about the error codes in the readme.md file.
* Imported esm globally in the main.js file aswell.
* More keywords in the package.json file.

## [**1.3.0**] - 2019-12-19

### Added

* More information about the author in the package.json file.
* Funding link in the package.json file.
* Errors are now controlled in the webpage.
* The webpage will show the number of html/php files found.
* The webpage will show the number of css files found.
* The webpage will show the number of ids found.
* The webpage will show the number of classes found.
* Better readability at the webpage.
* Better readability in the main.js file.

### Changed

* The CLI command will not show the files that doesn't contain CSS styles (empty).
* The CLI command will not show the number of html/php files found.
* The CLI command will not show the number of css files found.
* The CLI command will not show the number of ids found.
* The CLI command will not show the number of classes found.
* Image of the webpage in the readme.md file.
* Updated dependencies.

### Fixed

* If the server finds an error, it will now show an error, as before wasn't showing the error.
* If the server finds an error, it will try to close it in an expected way before showing the error.
* The extension will only show if there is a new version, if the user invokes the command implodeCss.
* Some typos.

### Removed

* Upcoming features in the readme.md file.
* Chalk import in the src/controller/getCssUnused.js file.

## [**1.2.0**] - 2019-12-13

### Added

* Result's page now it will also show the number of empty CSS files and their paths.

### Fixed

* Npm run tests, now it only triggers the cliTest file.

### Changed

* Audit and Fix code from the main.js file will receive an object of options instead of the options itself.
* Look of the headers of the result's page.
* Image of the result's page in the readme.md file.

## [**1.1.3**] - 2019-12-12

### Fixed

* Bug that stopped from showing the content.

## [**1.1.2**] - 2019-12-10

### Added

* Update-notifier, to see when a new version is out.

### Fixed

* Removed URLS when finding styles to avoid false positives.

## [**1.1.1**] - 2019-12-10

### Fixed

* Minor typos.
* Files without any css/ids (inconsistency) are now fixed and now it doesn't parse false positives.

### Removed

* Removed folders .vscode and node_modules from Github.

### Added

* More tests.
* Folder 'projectExampleForTests' inside the folder 'tests', that will contain an example project in order to do tests in it.

## [**1.1.0**] - 2019-12-07

### Added

* Now unexpected or unhandled errors will be controlled, and it will not throw errors.

## [**1.0.4**] - 2019-12-07

### Fixed

* Show information threw an error.

### Added

* NPM badges to the readme.md file.
* More error information in multiple catches.
* A .gitignore file with node_modules in it.

### Removed

* node_modules from the Github, as adding it to the .gitignore file.

### Changed

* If the user used unexpected arguments, it will show the manual and then the error.

## [**1.0.3**] - 2019-12-07

### Fixed

* Some typographical errors.

### Added

* More keywords to the package.json file.
* Documentation of the cliTest.js file.
* When fixing the code, now it will remove empty lines and lines with only '{}'.

### Changed

* Description of the package.
* Minor text changes in the readme.md file.

## [**1.0.2**] - 2019-12-06

### Added

* More screenshots to the readme.md file.

### Fixed

* Help command showed the version twice.

## [**1.0.1**] - 2019-12-06

### Added

* Basic readme documentation.

## [**1.0.0**] - 2019-12-06 **PUBLISH DATE**


src/exitCodes.js
/**
 * @file exitCodes file that contains a function that shows information about
 * the exit code and simply stops the process of the program.
 *
 * @author Jose Gracia Berenguer
 * @since 1.0.0
 * @link https://github.com/Josee9988/Implode-CSS
 */

import chalk from 'chalk';


/**
 * Summary: showCodeAndExit is a function that receives an error code
 * and it closes the process/program and shows a brief output of
 * what happened, and the error code.
 *
 * @param {number} code id of the exit code, that will be used in the switch.
 * @param {any} extraInformation object, array or string that may be used
 * or not if the error code needs to output extra information in order to
 * make the error more understandable.
 * @return {void}
 */
export function showCodeAndExit(code, extraInformation = 'Not given') {
    switch (code) {
        case 200: // exiting OK
            console.log(`Exiting without errors, thanks for using ${chalk.green('implodeCss')}`);
            break;

        case 201: // exiting OK? This exit is a default one, and should not be used.
            console.log(`Exiting without an expected code, thanks for using ${chalk.green('implodeCss')}`);
            break;

        case 400: // bad arguments passed
            // we remove the two first arguments: path to node and call to the package
            extraInformation.splice(0, 2);
            console.error(`${chalk.bold.red('ERROR: ')}your arguments: '${chalk.bold.magenta(extraInformation)}' are wrong.`);
            console.error(`${chalk.bold.red('errno:') + chalk.bold(' 400')} - Bad or unexpected arguments.`);
            console.error(`Check the manual by typing: ${chalk.bold.grey('implodeCss -h')}`);
            console.error(`If you believe this is our fault, please let us know at: ${chalk.bold('https://github.com/Josee9988/Implode-CSS/issues')}`);
            break;

        case 401: // path does not exist
            console.error(`${chalk.bold.red('ERROR: ')}your path: '${chalk.bold.magenta(extraInformation)}' is wrong.`);
            console.error(`${chalk.bold.red('errno: ') + chalk.bold(' 401')} - The given path does not exist.`);
            console.error(`If you beleive this is our fault, please let us know at: ${
                chalk.bold('https://github.com/Josee9988/Implode-CSS/issues')}`);
            break;

        case 402: // path don't have read permissions
            console.error(`${chalk.bold.red('ERROR: ')}your path: '${chalk.bold.magenta(extraInformation)}' is wrong.`);
            console.error(`${chalk.bold.red('errno: ') + chalk.bold(' 402')} - The given path does not have read permissions.`);
            console.error(`If you beleive this is our fault, please let us know at: ${
                chalk.bold('https://github.com/Josee9988/Implode-CSS/issues')}`);
            break;

        case 403: // path don't have write permissions
            console.error(`${chalk.bold.red('ERROR: ')}your path: '${chalk.bold.magenta(extraInformation)}' is wrong.`);
            console.error(`${chalk.bold.red('errno: ') + chalk.bold(' 403')} - The given path does not have write permissions.`);
            console.error(`If you beleive this is our fault, please let us know at: ${
                chalk.bold('https://github.com/Josee9988/Implode-CSS/issues')}`);
            break;

        default: // unknown error
            console.error(chalk.bold.red('Unknown error.'));
            console.error(`${chalk.bold.red('errno:') + chalk.bold(' 500')
                } - Unexpected error code, you should never be reading this, please inform us how this happened at: ${
                chalk.bold('https://github.com/Josee9988/Implode-CSS/issues')}`);
            process.exit('500');
            break;
    }
    process.exit(code); // exists with the code given
}

export default showCodeAndExit;
